### PR TITLE
Handle ambiguities introduced by stack association

### PR DESCRIPTION
**Pull request by idoru**
- update-buildpack/delete-buildpack now fail if the specified buildpack name is ambiguous
- update-buildpack/delete-buildpack now accept `[-s STACK]` flag to disambiguate
- create-buildpack removes stray buildpack records on buildpack-bits upload failure, which will eventually return errors when uploaded buildpack bits specify a stack which is not unique in combination with the buildpack name.

[#153256959]

Signed-off-by: Sam Coward <scoward@pivotal.io>
Signed-off-by: Jackson Feeny <jacksonfeeny@gmail.com>
Signed-off-by: Victoria Henry <vhenry@pivotal.io>

## What Need Does It Address?
In light of this [cloud_controller_ng PR](https://github.com/cloudfoundry/cloud_controller_ng/pull/1064), future versions of CC will support multiple buildpacks with the same name as long as they have different stacks. In the lens of this change, this affects the following buildpack CLI commands:

**update-buildpack/delete-buildpack**: these operate on a specified buildpack (e.g. update-buildpack/delete-buildpack) because these commands currently accept the buildpack name as the sole means of indicating the specific buildpack.

This change will allow specifying the stack in addition to the buildpack's name with the `-s` flag, but will only be required if the name is not distinct.

**create-buildpack**: Since the stack associated with buildpacks is only known by CC after buildpack bits are uploaded, it's possible that the API will fail on the buildpack bits upload if it is discovered that the stack + buildpack name is not unique. This leaves a stray buildpack entry with a stack of `unknown` which requires manual intervention to delete.

This change will make the CLI clean up the stray record.

## Possible Drawbacks

Unaware.

## Why Should This Be In Core?

It's related to core commands (*-buildpack)

## Description of the Change

- `BuildpackRepository` `findByName` now returns `AmbiguousModelError` if it finds more than 1
- `BuildpackRepository` now can `findByNameAndStack`
- `Update/DeleteBuildpack` commands use `findByNameAndStack` with stack passed via `-s` flag
- `Update/DeleteBuildpack` commands fail if `findByName` returns `AmbiguousModelError`
- `CreateBuildpack` command now deletes buildpack if the upload fails

## Alternate Designs

It's worth noting that `UpdateBuildpack` and `DeleteBuildpack` commands both have need to get a `Buildpack`, and that they do so by means of the specified buildpack name (and now, possible `-s` flag argument). They do this through different means; `DeleteBuildpack` interacts with `BuildpackRepository` directly and `UpdateBuildpack` is provided one via a `BuildpackRequirement`.

We initially sought to unify their use of a `BuildpackRequirement`, however we found that the `Requirements` mechanism takes away control of handling any errors generated by the requirement (they bubble up and always manifest as a command failure). 
 
This seemed in opposition to what appears to be intentional effort in the DeleteBuildpack command to handle this and exit successfully with a warning instead. In light of this and not wanting to introduce any widespread architectural changes so we have not done said unification and have ensured the exit code behavior is not modified.

## Applicable Issues
This change is part of a larger buildpacks story:
https://www.pivotaltracker.com/story/show/153256959

Other open PRs related to this:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1064
https://github.com/cloudfoundry/capi-release/pull/75

### CODE REVIEW

File: cf/commands/buildpack/delete_buildpack.go [hunk 1]
```diff
@@ -68,15 +78,22 @@ func (cmd *DeleteBuildpack) Execute(c flags.FlagContext) error {
...
+	} else {
+		buildpack, err = cmd.buildpackRepo.FindByNameAndStack(buildpackName, stack)
+		cmd.ui.Say(T("Deleting buildpack {{.BuildpackName}} with stack {{.Stack}}...", map[string]interface{}{"BuildpackName": terminal.EntityNameColor(buildpackName), "Stack": terminal.EntityNameColor(stack)}))
+	}
 
 	switch err.(type) {
 	case nil: //do nothing
 	case *errors.ModelNotFoundError:
 		cmd.ui.Ok()
 		cmd.ui.Warn(T("Buildpack {{.BuildpackName}} does not exist.", map[string]interface{}{"BuildpackName": buildpackName}))
```
Comment (tjvman) [inline cf/commands/buildpack/delete_buildpack.go]: Update this error message to include the stack if it is provided; there may be a buildpack with the given name but a different stack.

File: integration/isolated/delete_buildpack_command_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			var err error
+			dir, err = ioutil.TempDir("", "delete-buildpack-test")
+			Expect(err).ToNot(HaveOccurred())
+
+			filename := "manifest.yml"
+			manifestFile := filepath.Join(dir, filename)
+			err = ioutil.WriteFile(manifestFile, []byte("---\nstack:cflinuxfs2"), 0400)
+			Expect(err).ToNot(HaveOccurred())
+
+			session := CF("create-buildpack", "some-buildpack", dir, "1")
```
Comment (tjvman) [inline integration/isolated/delete_buildpack_command_test.go]: Looks like only one buildpack is being created here; this test should not depend on the behavior of `update-buildpack` integration test creating a buildpack with the same name.

File: integration/isolated/update_buildpack_command_test.go [hunk 3]
```diff
@@ -40,6 +40,38 @@ var _ = Describe("update-buildpack command", func() {
 		})
```
Comment (tjvman) [inline integration/isolated/update_buildpack_command_test.go]: Please add a test for updating a buildpack with a manifest that specifies a different stack. Per the original proposal, 
>  If the manifest.yml specifies a different stack, this is an error.
and there should be test ensuring that the error is bubbled up from the API.

**Comment by sclevine** [comment]
Would it make sense for us to make some of these changes (ex. `rename-buildpack` and better error messages) in a second PR, given that the current PR strictly improves the behavior with the latest version of CAPI?

It seems like we're leaving more stuff broken for longer by blocking this PR on additional functionality.

**Comment by tjvman** [comment]
@sclevine For adding `-s` to `rename-buildpack` and updating the text of the error messages, we're okay with a new PR. 

For the other issues, we would like to have the changes addressed in order to merge.

I also updated my review comment to include an additional issue - the PR changes the idempotency of `create-buildpack`; that will also need to be fixed before we can merge this PR, as it is a breaking change.

File: cf/api/buildpacks_test.go [hunk 4]
```diff
@@ -154,6 +154,90 @@ var _ = Describe("Buildpacks repo", func() {
...
+							  "position": 11
+						  }
+					  }
+					  ]
+				  }`},
+			}))
+
+			_, apiErr := repo.FindByName("Buildpack1")
+			Expect(handler).To(HaveAllRequestsCalled())
+			Expect(apiErr.(*errors.AmbiguousModelError)).NotTo(BeNil())
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/api/buildpacks_test.go]: 👍 Switched to using `BeAssignableToTypeOf` to test exactly what we were already doing, but without panicking.

File: cf/commands/buildpack/buildpacks.go [hunk 5]
```diff
@@ -58,7 +58,7 @@ func (cmd *ListBuildpacks) SetDependency(deps commandregistry.Dependency, plugin
 func (cmd *ListBuildpacks) Execute(c flags.FlagContext) error {
 	cmd.ui.Say(T("Getting buildpacks...\n"))
 
-	table := cmd.ui.Table([]string{"buildpack", T("position"), T("enabled"), T("locked"), T("filename")})
+	table := cmd.ui.Table([]string{"buildpack", T("stack"), T("position"), T("enabled"), T("locked"), T("filename")})
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/commands/buildpack/buildpacks.go]: 👍

File: cf/commands/buildpack/create_buildpack.go [hunk 6]
```diff
@@ -94,6 +94,7 @@ func (cmd *CreateBuildpack) Execute(c flags.FlagContext) error {
 
 	err = cmd.buildpackBitsRepo.UploadBuildpack(buildpack, buildpackFile, buildpackFileName)
 	if err != nil {
+		cmd.buildpackRepo.Delete(buildpack.GUID)
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/commands/buildpack/create_buildpack.go]: :+1: and fixed `create_buildpack_test` to match.

File: cf/commands/buildpack/update_buildpack_test.go [hunk 7]
```diff
@@ -112,7 +122,22 @@ var _ = Describe("Updating buildpack command", func() {
 		})
 	})
 
-	Context("updates buildpack when passed the proper flags", func() {
+	Describe("flags", func() {
+		Context("stack flag", func() {
+			It("updates the specific buildpack by name and stack, when stack is provided", func() {
+				runCommand("-i", "999", buildpackName, "-s", "cflinuxfs99")
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/commands/buildpack/update_buildpack_test.go]: I believe that it's not that we're testing `-i` here, but that we need to make a change to the buildpack in order to test update.

File: integration/isolated/delete_buildpack_command_test.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+		AfterEach(func() {
+			Expect(os.RemoveAll(dir)).To(Succeed())
+		})
+
+		It("accepts stack argument and deletes the buildpack", func() {
+			session := CF("delete-buildpack", "some-buildpack", "-s", "cflinuxfs2", "-f")
+			Eventually(session).Should(Exit(0))
+			Eventually(session.Out).Should(Say("OK"))
+			Expect(session.Err).NotTo(Say("Incorrect Usage:"))
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply integration/isolated/delete_buildpack_command_test.go]: 👍

File: integration/isolated/update_buildpack_command_test.go [hunk 9]
```diff
@@ -60,8 +92,8 @@ var _ = Describe("update-buildpack command", func() {
...
 
 		AfterEach(func() {
-			err := os.RemoveAll(dir)
-			Expect(err).NotTo(HaveOccurred())
+			Eventually(CF("delete-buildpack", "some-buildpack", "-f")).Should(Exit(0))
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply integration/isolated/update_buildpack_command_test.go]: 👍

File: cf/api/apifakes/old_fake_buildpack_repo.go [hunk 10]
```diff
@@ -42,17 +50,34 @@ func (repo *OldFakeBuildpackRepository) FindByName(name string) (buildpack model
...
+	buildpack = repo.FindByNameAndStackBuildpack
+
+	if repo.FindByNameAndStackNotFound {
+		apiErr = errors.NewModelNotFoundError("Buildpack", name)
 	}
 
 	return
 }
 
+var buildpackCreateCount int
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/api/apifakes/old_fake_buildpack_repo.go]: 👍 All buildpacks this creates will now have the same GUID.

File: cf/errors/ambiguous_model_error.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+func NewAmbiguousModelError(modelType, name string) error {
+	return &AmbiguousModelError{
+		ModelType: modelType,
+		ModelName: name,
+	}
+}
+
+func (err *AmbiguousModelError) Error() string {
```
[the remark this answers is not in the document]
Comment (astrieanna) [reply cf/errors/ambiguous_model_error.go]: ⏭ will update error messages in next PR.

File: cf/commands/buildpack/delete_buildpack.go (hunk 1 above)
Comment (astrieanna) [reply cf/commands/buildpack/delete_buildpack.go]: ⏭will update error messages in next PR.

File: integration/isolated/delete_buildpack_command_test.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
+package isolated
```
[the remark this answers is not in the document]
Comment (djoyahoy) [reply integration/isolated/delete_buildpack_command_test.go]: 👍

File: integration/isolated/update_buildpack_command_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djoyahoy) [reply integration/isolated/update_buildpack_command_test.go]: 👍

Comment (djoyahoy) [reply integration/isolated/update_buildpack_command_test.go]: 👍

File: integration/isolated/delete_buildpack_command_test.go (hunk 2 above)
Comment (jfeeny) [reply integration/isolated/delete_buildpack_command_test.go]: 👍

**Comment by sclevine** [comment]
@abbyachau @tjvman ready to merge

**Comment by astrieanna** [comment]
@tjvman I think we've made all the changes requested for this PR.

**Comment by anyu** [comment]
Apart from the ongoing `delete-buildpack` updates, the changes in [this commit](https://github.com/cloudfoundry/cli/pull/1320/commits/8402d2a69302eef73572bed98e92bfde5c125a21) look good.

However, we want to point out that the behavior around attempting to create a buildpack with the same stack when it already exists is weird/pretty confusing, even if it's technically correct.

When attempting to create a buildpack / stack combination that already exists, we see the following warning the first time around:
```
± |idoru-buildpack-stack-assoc {2} U:2 ✗| → cf create-buildpack test /tmp/test-buildpack.zip 10
Creating buildpack test...
OK

Uploading buildpack test...
Done uploading               
OK
Server error, status code: 422, error code: 290000, message: The buildpack name test is already in use for the stack cflinuxfs2
TIP: use 'cf update-buildpack' to update this buildpack
```
This exits 0, which is the expected behavior (though we prefer that the CC error be wrapped in a nicer warning message without `server error`). 

When attempting to create a buildpack / stack combination that already exists a second time however, we see the following warning:
```
± |idoru-buildpack-stack-assoc {2} U:2 ✗| → cf create-buildpack test /tmp/test-buildpack.zip 10
Creating buildpack test...
OK
Buildpack test already exists without a stack
TIP: use 'cf delete-buildpack' to delete buildpack test without a stack
```
This happens because creating a buildpack with a stack is a two step process - per [the doc](https://docs.google.com/document/d/1hPvZuNsqk08ysw1wvmRkkqbr-6PFRxgLVgYEchf38z0/edit#heading=h.i8llyd6xqjwo), when buildpack records are first created the stack will be nil. Per the original PR comment, "the API will fail on the buildpack bits upload if it is discovered that the stack + buildpack name is not unique. This leaves a stray buildpack entry with a stack of unknown which requires manual intervention to delete." 

It sounds like our team had pushed back on the part where we clean up the stray record, so technically it's true that "Buildpack test already exists without a stack'. But to a user, this message is unclear because they've been attempting to create a buildpack with a stack name this whole time.

Not sure if there's any action to be taken, but just want to call that out for discussion if needed.

cc @abbyachau

**Comment by sclevine** [comment]
We agree that in the case where `create-buildpack` fails after creating the buildpack record, the best user experience would be for the CLI to remove the stray buildpack record. We had implemented this in the PR, but we were told to remove it. Happy to restore it if you'd prefer.

**Comment by astrieanna** [comment]
We have implemented the new delete behavior and added/corrected unit and integration tests for it.

**Pull request merged**